testing/run_ap_with_constraint.py

```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/python3
+
+# This script run vpr with fixed io constraints.
+
+import argparse
+import os
+import shutil
+import subprocess
+import sys
+
+# This function parses arguments of this python script
+def command_parser():
+    parser = argparse.ArgumentParser(description="parse arguments for fix io script")
+    parser.add_argument("test_suite_name", help="name of the one test suite being run")
+    default_vtr_path = os.path.join(os.path.expanduser("~"), "vtr-verilog-to-routing-ap")
+    parser.add_argument("-vtr_path", default=default_vtr_path, type=str, help="path of root directory for VTR")
+    default_test_cases_path = os.path.join(os.getcwd(), "tests")
+    parser.add_argument("-test_cases_path", default=default_test_cases_path, type=str, help="path of to intermediate files of the testcases")
+    parser.add_argument("-chan_width", default=100, type=int, help="largest chan width required for tests in this test suite")
+
+    return parser
+    
+# This is the main function of the python script
+def run_test_main(args):
+    # parse arguments
+    args = command_parser().parse_args(args)
+
+    # check if test suite and test directory exist. 
+    test_path = os.path.join(args.test_cases_path, args.test_suite_name)
+    if(not os.path.isdir(test_path)):
+        print(test_path, " does not exists!")
+
+    # iterate through arch then circuit and call vpr on it
+    # results are in tests/arch/circuit/common/ap_run
+    for arch in os.listdir(test_path):
+        arch_dir_path = os.path.join(test_path, arch)
+        for circuit in os.listdir(arch_dir_path):
+            circuit_dir_path = os.path.join(arch_dir_path, circuit, "common")
+            ap_run_dir = os.path.join(circuit_dir_path, "ap_dir")
+            os.makedirs(ap_run_dir, exist_ok=True)
+            script_path = os.getcwd();
+            os.chdir(ap_run_dir)
```

Comment:
I would be careful with running VPR tests within the same directory. Naming conventions may coincide and cause some very weird affects. I would look at my routing testing script and see what I did. I recommend creating a new folder for each run (run001, run002, etc.) and then giving each circuit its own folder.

See:
https://github.com/AlexandreSinger/fine-grained-parallel-router/blob/main/testing/run_test.py